File: src/tqi/scoring/tcqsm.py
"""TCQSM Level of Service computation per route.

Reference: Transit Capacity and Quality of Service Manual, 3rd Edition
(TCRP Report 165, Transportation Research Board).
"""

from dataclasses import dataclass

import numpy as np
import pandas as pd

from tqi.config import TCQSM_LOS
from tqi.gtfs.parse import GTFSFeed


@dataclass
class RouteLOS:
    route_name: str
    route_long_name: str
    route_id: str
    n_trips: int
    median_headway_min: float
    peak_headway_min: float | None  # AM peak 7-9
    los_grade: str
    los_description: str


def _headway_to_los(headway_min: float) -> tuple[str, str]:
    """Map median headway to TCQSM LOS grade."""
    for max_hw, grade, desc in TCQSM_LOS:
        if headway_min <= max_hw:
            return grade, desc
    return "F", "Service unattractive to all riders"


def compute_route_los(feed: GTFSFeed) -> list[RouteLOS]:
    """Compute TCQSM LOS for each route in the filtered feed."""
    route_cols = ["route_id", "route_short_name"]
    if "route_long_name" in feed.routes.columns:
        route_cols.append("route_long_name")

    st = feed.stop_times.merge(feed.trips[["trip_id", "route_id", "direction_id"]], on="trip_id")
    st = st.merge(feed.routes[route_cols], on="route_id")

    # Build route_id → long name lookup
    long_names: dict[str, str] = {}
    if "route_long_name" in feed.routes.columns:
        for _, row in feed.routes.iterrows():
            long_names[str(row["route_id"])] = str(row.get("route_long_name", ""))

    results = []

    for (route_id, route_name), group in st.groupby(["route_id", "route_short_name"]):
        # Compute headway per direction separately, then take the better one.
        # Mixing directions inflates frequency (two 30-min routes interleave to look like 15-min).
        n_trips = group["trip_id"].nunique()
        dir_headways = []
        for d_id, d_group in group.groupby("direction_id"):
            d_starts = (
                d_group.sort_values("stop_sequence")
                .groupby("trip_id")["departure_min"]
                .first()
                .sort_values()
            )
            if len(d_starts) >= 2:
                hws = np.diff(d_starts.values.astype(float))
                hws = hws[hws > 0]
                if len(hws) > 0:
                    dir_headways.append(float(np.median(hws)))
        # If we have per-direction data, use the best (lowest) direction headway.
        # If only one direction exists, use that.
        if dir_headways:
            trip_starts_median = min(dir_headways)
        else:
            # Fallback: all trips combined
            all_starts = (
                group.sort_values("stop_sequence")
                .groupby("trip_id")["departure_min"]
                .first()
                .sort_values()
            )
            hws = np.diff(all_starts.values.astype(float))
            hws = hws[hws > 0]
            trip_starts_median = float(np.median(hws)) if len(hws) > 0 else 999.0

        if n_trips < 2:
            results.append(RouteLOS(
                route_name=str(route_name),
                route_long_name=long_names.get(str(route_id), ""),
                route_id=str(route_id),
                n_trips=n_trips,
                median_headway_min=999.0,
                peak_headway_min=None,
                los_grade="F",
                los_description="Service unattractive to all riders",
            ))
            continue

        median_hw = trip_starts_median

        # AM peak headway per direction (7:00-9:00 = 420-540 min)
        peak_hw = None
        for d_id, d_group in group.groupby("direction_id"):
            d_starts = (
                d_group.sort_values("stop_sequence")
                .groupby("trip_id")["departure_min"]
                .first()
                .sort_values()
            ).values.astype(float)
            peak_s = d_starts[(d_starts >= 420) & (d_starts <= 540)]
            if len(peak_s) >= 2:
                phw = np.diff(peak_s)
                phw = phw[phw > 0]
                if len(phw) > 0:
                    candidate = float(np.median(phw))
                    if peak_hw is None or candidate < peak_hw:
                        peak_hw = candidate

        grade, desc = _headway_to_los(median_hw)

        results.append(RouteLOS(
            route_name=str(route_name),
            route_long_name=long_names.get(str(route_id), ""),
            route_id=str(route_id),
            n_trips=n_trips,
            median_headway_min=round(median_hw, 1),
            peak_headway_min=round(peak_hw, 1) if peak_hw else None,
            los_grade=grade,
            los_description=desc,
        ))

    # Sort by route name (numeric sort where possible)
    results.sort(key=lambda r: (r.route_name.zfill(5)))
    return results


def compute_system_los_summary(route_los: list[RouteLOS]) -> dict:
    """Aggregate system-level LOS statistics."""
    grades = [r.los_grade for r in route_los]
    headways = [r.median_headway_min for r in route_los if r.median_headway_min < 999]

    grade_counts = {}
    for g in ["A", "B", "C", "D", "E", "F"]:
        grade_counts[g] = grades.count(g)

    return {
        "n_routes": len(route_los),
        "grade_counts": grade_counts,
        "median_system_headway_min": float(np.median(headways)) if headways else 0,
        "best_grade": min(grades) if grades else "F",
        "worst_grade": max(grades) if grades else "F",
        "pct_los_d_or_worse": sum(1 for g in grades if g >= "D") / len(grades) * 100 if grades else 0,
    }

File: go/internal/scoring/tcqsm.go
package scoring

import (
	"math"
	"sort"

	"github.com/jasondyck/chwk-tqi/internal/config"
	"github.com/jasondyck/chwk-tqi/internal/gtfs"
)

// RouteLOS holds the level-of-service grading for a single route.
type RouteLOS struct {
	RouteName     string   `json:"route_name"`
	RouteLongName string   `json:"route_long_name"`
	RouteID       string   `json:"route_id"`
	NTrips        int      `json:"n_trips"`
	MedianHeadway float64  `json:"median_headway"`
	PeakHeadway   *float64 `json:"peak_headway"`
	LOSGrade      string   `json:"los_grade"`
	LOSDescription string  `json:"los_description"`
}

// SystemLOSSummary summarises the LOS across all routes.
type SystemLOSSummary struct {
	NRoutes             int                `json:"n_routes"`
	GradeCounts         map[string]int     `json:"grade_counts"`
	MedianSystemHeadway float64            `json:"median_system_headway"`
	BestGrade           string             `json:"best_grade"`
	WorstGrade          string             `json:"worst_grade"`
	PctLOSDOrWorse      float64            `json:"pct_los_d_or_worse"`
}

// HeadwayToLOS maps a headway (minutes) to a TCQSM grade and description.
func HeadwayToLOS(headwayMin float64) (string, string) {
	for _, g := range config.TCQSMLOS {
		if headwayMin <= float64(g.MaxHeadwayMin) {
			return g.Grade, g.Description
		}
	}
	return "F", "Service unattractive to all riders"
}

// ComputeRouteLOS grades each route in the feed by its median headway
// per direction.
func ComputeRouteLOS(feed *gtfs.Feed) []RouteLOS {
	// Build lookup maps.
	tripByID := make(map[string]*gtfs.Trip, len(feed.Trips))
	for i := range feed.Trips {
		tripByID[feed.Trips[i].TripID] = &feed.Trips[i]
	}
	routeByID := make(map[string]*gtfs.Route, len(feed.Routes))
	for i := range feed.Routes {
		routeByID[feed.Routes[i].RouteID] = &feed.Routes[i]
	}

	// Collect first departure per trip.
	tripFirstDep := make(map[string]int)
	for _, st := range feed.StopTimes {
		if st.StopSequence == 1 {
			tripFirstDep[st.TripID] = st.DepartureMin
		} else if cur, ok := tripFirstDep[st.TripID]; !ok || st.DepartureMin < cur {
			tripFirstDep[st.TripID] = st.DepartureMin
		}
	}

	// Group trips by (routeID, directionID).
	type dirKey struct{ routeID, dir string }
	dirTrips := make(map[dirKey][]int) // departure minutes
	for tripID, dep := range tripFirstDep {
		tr, ok := tripByID[tripID]
		if !ok {
			continue
		}
		k := dirKey{tr.RouteID, tr.DirectionID}
		dirTrips[k] = append(dirTrips[k], dep)
	}

	// Compute headways per direction, take median.
	type routeResult struct {
		routeID   string
		nTrips    int
		headways  []float64
	}
	routeResults := make(map[string]*routeResult)

	for k, deps := range dirTrips {
		sort.Ints(deps)
		for i := 1; i < len(deps); i++ {
			hw := float64(deps[i] - deps[i-1])
			if hw <= 0 {
				continue
			}
			rr, ok := routeResults[k.routeID]
			if !ok {
				rr = &routeResult{routeID: k.routeID}
				routeResults[k.routeID] = rr
			}
			rr.headways = append(rr.headways, hw)
		}
		if rr, ok := routeResults[k.routeID]; ok {
			rr.nTrips += len(deps)
		} else {
			routeResults[k.routeID] = &routeResult{routeID: k.routeID, nTrips: len(deps)}
		}
	}

	var results []RouteLOS
	for routeID, rr := range routeResults {
		rt := routeByID[routeID]
		if rt == nil {
			continue
		}

		var medHW float64
		if len(rr.headways) > 0 {
			sort.Float64s(rr.headways)
			medHW = median(rr.headways)
		} else {
			medHW = math.Inf(1)
		}

		grade, desc := HeadwayToLOS(medHW)
		results = append(results, RouteLOS{
			RouteName:      rt.RouteShortName,
			RouteLongName:  rt.RouteLongName,
			RouteID:        routeID,
			NTrips:         rr.nTrips,
			MedianHeadway:  medHW,
			LOSGrade:       grade,
			LOSDescription: desc,
		})
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].RouteName < results[j].RouteName
	})
	return results
}

// ComputeSystemLOSSummary aggregates route-level LOS into a system summary.
func ComputeSystemLOSSummary(routeLOS []RouteLOS) *SystemLOSSummary {
	summary := &SystemLOSSummary{
		NRoutes:     len(routeLOS),
		GradeCounts: make(map[string]int),
	}
	if len(routeLOS) == 0 {
		return summary
	}

	gradeOrder := map[string]int{"A": 0, "B": 1, "C": 2, "D": 3, "E": 4, "F": 5}
	bestIdx := 5
	worstIdx := 0

	var headways []float64
	var dOrWorse int

	for _, r := range routeLOS {
		summary.GradeCounts[r.LOSGrade]++
		if !math.IsInf(r.MedianHeadway, 1) {
			headways = append(headways, r.MedianHeadway)
		}
		if idx, ok := gradeOrder[r.LOSGrade]; ok {
			if idx < bestIdx {
				bestIdx = idx
			}
			if idx > worstIdx {
				worstIdx = idx
			}
			if idx >= 3 { // D=3, E=4, F=5
				dOrWorse++
			}
		}
	}

	gradeNames := []string{"A", "B", "C", "D", "E", "F"}
	summary.BestGrade = gradeNames[bestIdx]
	summary.WorstGrade = gradeNames[worstIdx]

	if len(headways) > 0 {
		sort.Float64s(headways)
		summary.MedianSystemHeadway = median(headways)
	}

	summary.PctLOSDOrWorse = float64(dOrWorse) / float64(len(routeLOS)) * 100.0

	return summary
}

func median(sorted []float64) float64 {
	n := len(sorted)
	if n == 0 {
		return 0
	}
	if n%2 == 0 {
		return (sorted[n/2-1] + sorted[n/2]) / 2.0
	}
	return sorted[n/2]
}
